WIP: Add RE2 regexp engine support

WGH- opened the pull request:
See #1996.

Every commit has detailed description.

I've implemented the fallback approach so far: if RE2 support is enabled, every regular expression is compiled with RE2, and if that fails, regular expression is compiled with libpcre instead. Please see #1996 for more detailed discussion why this approach is necessary.

There's also lots of refactoring of regexp related code.

Open questions:
- [ ] Perhaps, since RE2 support is considered experimental, configure script should not even attempt to build it unless explicit `--with-re` is supplied?
- [ ] There's no easy way to know which regular expression uses which engine. It might feel like  kind of implicit magic, which isn't really good.
- [x] There's no error handling when regular expression is invalid. This isn't a regression introduced by this PR, but now might be the time to do something about it.
- [ ] PR needs code review, obviously.
- [ ] Changelog entry
- [ ] Documentation
- [ ] Maybe something else I missed.

zimmerle committed:
Moving regex from utils to its own namespace

zimmerle committed:
Renames SMatch to RegexMatch

zimmerle committed:
Removes unecessary static methods from regex class

zimmerle committed:
regex: Organizes the classes per file

moves Pcre to backends/pcre.cc
moves PegexMatch to regex_match.h

zimmerle committed:
Makes re2 detectable by the build scripts

WGH- committed:
Add support for capturing group test cases

This enables unit tests to compare the matching groups as well,
not just binary match-no match.

WGH- committed:
Fix Pcre::searchAll behaviour wrt empty capturing groups

Previously, searchAll would stop search when it encountered an empty
matching group in any position. This means that, for example,
regular expression "(a)(b?)(c)" would match string "ac", but the
resulting group list would be ["ac", "a"].

After this change, the resulting list for the aforementioned regular
expression becomes ["ac", "a", "", "c"] like it should've been.

Additionally, this also changes behaviour for multiple matches. For
example, when "aaa00bbb" is matched by "[a-z]*", previously only "aaa"
would be returned. Now the matching list is ["aaa", "", "", "bbb", ""].

The old behaviour was confusing and almost certainly a bug. The new
behaviour is the same as in Python's re.findall.

For reference, though, Go does it somewhat differently: empty matches
at the end of non-empty matches are ignored, so in Go above example is
["aaa", "", "bbb"] instead.

WGH- committed:
Add RE2 regex backend

WGH- committed:
Enable RE2 testing with Travis

Ubuntu 14.04 doesn't have RE2 package altogether,
and Ubuntu 16.04 RE2 package is too old.

Ubuntu 18.04 RE2 package might work, but this Ubuntu verison it's not
supported by Travis yet.

So build RE2 from source.

WGH- committed:
Add proper error handling to @rx operator

WGH- commented:
I've just force-pushed updated PR. 

* New commit with fixed error handling for the rx operator.
* `get_pattern` was renamed to `getPattern` to adhere to naming convention used elsewhere in ModSecurity.
* verifyCC conversion to the new classes, which was done very sloppily earlier, has been re-done properly.

WGH- commented:
Since my changes have been essentially merged into `v3/dev/re2`, I updated the PR to target `v3/master` branch now.

zimmerle commented:
Overall, the performance for the libpcre users was not impacted by this pull request:

![screenshot_20190206_164323](https://user-images.githubusercontent.com/428268/52369025-833a2980-2a2e-11e9-9302-d46fc0e882cf.png)

The chart was generated using the code merged into v3/dev/re2.

WGH- commented:
Ping.

WGH- commented:
Just in case, even though it's marked WIP, it's more than ready for the first round of review.

WGH- committed:
Implement RE2 fallback to libpcre

RE2 doesn't support certain features, like negative lookaround,
so when a regular expression cannot be compiled with RE2, it's
compiled with libpcre instead.

This has some runtime cost, as this fallback is implemented
with an extra heap object and virtual function calls.

When RE2 is not enabled, however, everything works as it did before.

WGH- commented:
Rebased on top of the current master.

eddie4 commented:
Very interesting commit. We recently ran into performance issues with modsecurity. Any news on how much improvement we might see?

zimmerle commented:
> Very interesting commit. We recently ran into performance issues with modsecurity. Any news on how much improvement we might see?

Hi @eddie4,

That is not the single front that we have on the performance side. Check the branch v3/dev/3.1-experimental. We have had great progress there. This is still on the Queue, to be merged once other pieces of v3.1 are complete.

WGH- commented:
> Very interesting commit. We recently ran into performance issues with modsecurity. Any news on how much improvement we might see?

@eddie4 in my experience, it doesn't really make it much faster in average, however, it eliminates ReDoS worst cases.

WGH- commented:
@zimmerle please ping me when it's time to rebase this on top `v3/dev/3.1-experimental`.

zimmerle commented:
> @zimmerle please ping me when it's time to rebase this on top `v3/dev/3.1-experimental`.

That is definitively something that I want to see happening. Lets have 3.1-experimental stable first. Working on the new issues that you have reported on #2376

WGH- committed:
Fix pkg-config Libs.private wrt RE2

mathetake commented:
hi - sorry to ping on this really old patch but I wonder by any chance we could see this happening in the future ...

airween commented:
Hi @mathetake,

thanks for bringing this up.
As you can see  GH status, this PR is very old therefore there are many conflicts which needs to resolve.

I think the best would be that we add a new PR based on this one, but the base would be the current status.

I try to handle this task soon - or feel free to send a new PR.

mathetake commented:
cool, thank you for the response @airween - i won't be able to work on this personally anytime soon, so feel free to take it over. I would love to test it out, review etc
